pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -241,226 +255,768 @@ public TransformResultMetadata getResultMetadata() {
+        Pair<byte[][], RoaringBitmap> byteValuesNullPair =
+            transformFunction.transformToBytesValuesSVWithNull(valueBlock);
+        byte[][] byteValues = byteValuesNullPair.getLeft();
+        RoaringBitmap nullBitmap = byteValuesNullPair.getRight();
+        for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+          if (selected[docId] == i) {
+            _bytesValuesSV[docId] = byteValues[docId];
+            if (nullBitmap != null && nullBitmap.contains(docId)) {
+              bitmap.add(docId);
+            }
+            unselectedDocs.clear(docId);
+          }
+        }
+        if (unselectedDocs.isEmpty()) {
+          break;
+        }
+      }
+    }
+    if (!unselectedDocs.isEmpty()) {
+      if (_elseStatement == null) {
+        for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+          _bytesValuesSV[docId] = (byte[]) DataSchema.ColumnDataType.BYTES.getNullPlaceholder();
+          bitmap.add(docId);
+        }
+      } else {
+        Pair<byte[][], RoaringBitmap> byteValuesNullPair = _elseStatement.transformToBytesValuesSVWithNull(valueBlock);
+        byte[][] byteValues = byteValuesNullPair.getLeft();
+        RoaringBitmap nullBitmap = byteValuesNullPair.getRight();
+        for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+          _bytesValuesSV[docId] = byteValues[docId];
+          if (nullBitmap != null && nullBitmap.contains(docId)) {
+            bitmap.add(docId);
+          }
+        }
+      }
+    }
+    return ImmutablePair.of(_bytesValuesSV, bitmap);
+  }
+
+  @Override
+  public RoaringBitmap getNullBitmap(ValueBlock valueBlock) {
+    int[] selected = getSelectedArray(valueBlock, true);
+    int numDocs = valueBlock.getNumDocs();
+    int numThenStatements = _thenStatements.size();
+    BitSet unselectedDocs = new BitSet();
+    unselectedDocs.set(0, numDocs);
+    final RoaringBitmap bitmap = new RoaringBitmap();
+    for (int i = 0; i < numThenStatements; i++) {
+      if (_selections[i]) {
+        TransformFunction transformFunction = _thenStatements.get(i);
+        RoaringBitmap nullBitmap = transformFunction.getNullBitmap(valueBlock);
+        for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+          if (selected[docId] == i) {
+            if (nullBitmap != null && nullBitmap.contains(docId)) {
+              bitmap.add(docId);
+            }
+            unselectedDocs.clear(docId);
+          }
+        }
+        if (unselectedDocs.isEmpty()) {
+          break;
+        }
+      }
+      if (!unselectedDocs.isEmpty()) {
+        if (_elseStatement == null) {
+          for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+            bitmap.add(docId);
```

Comment:
We can replace this with bitmap.or(unselectedDocs)?

Reply:
We cannot because unselectedDocs is a bitset and bitmap is Roaringbitmap. :)